### PR TITLE
Feature/008-Bugfix&Refactoring -> dev

### PR DESCRIPTION
**Pull request by WMinowski**

### CODE REVIEW

File: MVC-App/Common/Attributes/MaxValueAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Web;
+
+namespace MVC_App.Common.Attributes
+{
+    public class MaxValueAttribute : ValidationAttribute
+    {
+
```
Comment (dmyhun-andersen) [inline MVC-App/Common/Attributes/MaxValueAttribute.cs]: remove 11 and 12 lines

File: MVC-App/Common/Attributes/MaxValueAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public MaxValueAttribute(int maxValue)
+        {
+            _maxValue = maxValue;
+        }
+        public override bool IsValid(object value)
+        {
+            if (value != null)
+            {
+                int appliedValue = int.Parse(value.ToString());
```
Comment (dmyhun-andersen) [inline MVC-App/Common/Attributes/MaxValueAttribute.cs]: What if value="Hello world"? 
You need to use TryParse

Comment (dmyhun-andersen) [reply MVC-App/Common/Attributes/MaxValueAttribute.cs]: The same for MinValueAttribute

File: MVC-App/Controllers/HomeController.cs [hunk 3]
```diff
@@ -9,7 +9,6 @@ namespace MVC_App.Controllers
 {
     public class HomeController : Controller
```
Comment (dmyhun-andersen) [inline MVC-App/Controllers/HomeController.cs]: Remove HomeController with home views if you dont use them

File: MVC-App/Controllers/RelationsController.cs [hunk 4]
```diff
@@ -8,163 +8,33 @@
...
-
-            tblRelation tblRelation = await db.tblRelation.FindAsync(id);
-
-            if (tblRelation == null)
-            {
-                return HttpNotFound();
-            }
-
-            return View(tblRelation);
+            return View(await relationService.GetAsync(categoryId));
```
Comment (dmyhun-andersen) [inline MVC-App/Controllers/RelationsController.cs]: I've already told you. It is a bad practise to put everything in one line. 
Separate it.

File: MVC-App/Controllers/RelationsController.cs [hunk 5]
```diff
@@ -8,163 +8,33 @@
...
         }
 
         // GET: tblRelations/Create
```
Comment (dmyhun-andersen) [inline MVC-App/Controllers/RelationsController.cs]: Are you sure GET: tblRelations/Create - is correct? GET: Relations/Create

File: MVC-App/Repositories/UnitOfWork.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+using System.Threading.Tasks;
+using System.Web;
+
+namespace MVC_App.Repositories
+{
+    public class UnitOfWork : IDisposable
+    {
+        private RelationContext context = new RelationContext();
+
+        private GenericRepository<Relation> relationRepository;
```
Comment (dmyhun-andersen) [inline MVC-App/Repositories/UnitOfWork.cs]: Please create RelationRepository
...
    public class RelationRepository : GenericRepository<Relation>, IRepository<Relation>
    {
    }
...
Then youe line 14 will be private RelationRepository relationRepository;

Comment (dmyhun-andersen) [reply MVC-App/Repositories/UnitOfWork.cs]: The same for the others repos.

File: MVC-App/Services/RelationService.cs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+            return relationModels.ToList();
+        }
+
+        public string ApplyMask(string value, string mask)
+        {
+            if (mask == string.Empty || mask == null || value == string.Empty) return value;
+
+            List<char> result = new List<char>();
+
+            int valueIterator = 0;
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs]: can you make:
int valueIterator, valueNumbersCount, maskNumbersCount, maskLettersCount = 0;

File: MVC-App/Services/RelationService.cs [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            }
+
+            if (valueNumbersCount != maskNumbersCount || valueLettersCount != maskLettersCount)
+            {
+                return value;
+            }
+
+            for (int i = 0; i < mask.Length; i++)
+            {
+
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs]: remove empty line

File: MVC-App/Services/RelationService.cs [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+            var relationListVM = new RelationListVM { RelationViewModels = relationModels, Categories = Categories, Countries = countries };
+
+            return relationListVM;
+        }
+
+        public async Task Create(CreateEditRelationVM relationVM)
+        {
+            var relation = new Relation
+            {
+                Id = Guid.NewGuid(),
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs]: You can check if can you use AutoMapper here to map models from CreateEditRelationVM to Relation
You can implement in in separate task

File: MVC-App/Common/Attributes/MaxValueAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            if (int.TryParse(value.ToString(), out appliedValue))
+            {
+                if (appliedValue <= _maxValue)
+                    return true;
+                else
+                    this.ErrorMessage = "Value should be less than " + _maxValue.ToString();
+            }
+            return false;
+        }
+
```
Comment (dmyhun-andersen) [inline MVC-App/Common/Attributes/MaxValueAttribute.cs]: Remove empty lines

File: MVC-App/Common/Attributes/MinValueAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            if (int.TryParse(value.ToString(), out appliedValue))
+            {
+                if (appliedValue >= _minValue)
+                    return true;
+                else
+                    this.ErrorMessage = "Value should be more than " + _minValue.ToString();
+            }
+            return false;
+        }
+
```
Comment (dmyhun-andersen) [inline MVC-App/Common/Attributes/MinValueAttribute.cs]: Remove empty line

File: MVC-App/Domain.Models/AddressType.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        public DateTime? Timestamp1 { get; set; }
+
+        public DateTime? Timestamp2 { get; set; }
+
+        public DateTime? Timestamp3 { get; set; }
+
+        public DateTime? Timestamp4 { get; set; }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly")]
+        public virtual ICollection<RelationAddress> tblRelationAddress { get; set; }
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/AddressType.cs]: tblRelationAddress  -> RelationAddresses

File: MVC-App/Domain.Models/Country.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        [StringLength(3)]
+        public string ISO3166_3 { get; set; }
+
+        public Guid? DefaultVatId { get; set; }
+
+        [StringLength(255)]
+        public string PostalCodeFormat { get; set; }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly")]
+        public virtual ICollection<RelationAddress> tblRelationAddress { get; set; }
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/Country.cs]: the same

File: MVC-App/Domain.Models/Relation.cs [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+
+        public Guid? ThirdPartyToUseForInvoicing { get; set; }
+
+        public int? Flags { get; set; }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly")]
+        public virtual ICollection<RelationAddress> tblRelationAddress { get; set; }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly")]
+        public virtual ICollection<RelationCategory> tblRelationCategory { get; set; }
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/Relation.cs]: the same

File: MVC-App/Domain.Models/RelationAddress.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+        public double? Latitude { get; set; }
+
+        public Guid Id { get; set; }
+
+        public virtual AddressType tblAddressType { get; set; }
+
+        public virtual Country tblCountry { get; set; }
+
+        public virtual Relation tblRelation { get; set; }
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/RelationAddress.cs]: the same

File: MVC-App/Domain.Models/RelationCategory.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public Guid RelationId { get; set; }
+
+        public Guid CategoryId { get; set; }
+
+        public Guid Id { get; set; }
+
+        public virtual Category tblCategory { get; set; }
+
+        public virtual Relation tblRelation { get; set; }
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/RelationCategory.cs:20]: the same

File: MVC-App/Repositories/GenericRepository.cs [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+using System.Linq.Expressions;
+using System.Threading.Tasks;
+using System.Web;
+using System.Web.Mvc;
+
+namespace MVC_App.Repositories
+{
+    public class GenericRepository<TEntity> : IRepository<TEntity> where TEntity : class
+    {
+        private RelationContext dbContext;
```
Comment (dmyhun-andersen) [inline MVC-App/Repositories/GenericRepository.cs]: Nooo.
In GenericRepository<TEntity> can't be RelationContext.

File: MVC-App/Repositories/GenericRepository.cs [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            dbSet.Remove(entityToDelete);
+        }
+        public virtual void Update(TEntity entityToUpdate)
+        {
+            dbSet.Attach(entityToUpdate);
+            dbContext.Entry(entityToUpdate).State = EntityState.Modified;
+        }
+
+        //public void Dispose()
```
Comment (dmyhun-andersen) [inline MVC-App/Repositories/GenericRepository.cs]: Remove it

File: MVC-App/Domain.Models/Relation.cs [hunk 19]
```diff
@@ -22,7 +22,7 @@ public Relation()
 
         [Required]
         [StringLength(50)]
-        public string CreatedBy { get; set; }
+        public string CreatedBy { get; set; } = "admin";
```
Comment (dmyhun-andersen) [inline MVC-App/Domain.Models/Relation.cs:25]: remove  = "admin";

File: MVC-App/Services/RelationService.cs [hunk 20]
```diff
@@ -75,15 +77,7 @@ public string ApplyMask(string value, string mask)
...
-            int valueIterator = 0;
-
-            int valueNumbersCount = 0;
-
-            int valueLettersCount = 0;
-
-            int maskNumbersCount = 0;
-
-            int maskLettersCount = 0;
+            int valueIterator = 0, valueNumbersCount = 0, valueLettersCount = 0, maskNumbersCount = 0, maskLettersCount = 0;
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs:80]: int valueIterator, valueNumbersCount, valueLettersCount, maskNumbersCount, maskLettersCount = 0;

File: MVC-App/Services/RelationService.cs [hunk 21]
```diff
@@ -200,36 +193,40 @@ public async Task<RelationListVM> GetAsync(Guid? categoryId)
...
-                IsTemporary = false,
-                IsMe = false,
-                PaymentViaAutomaticDebit = false,
-                InvoiceDateGenerationOptions = 0,
-                InvoiceGroupByOptions = 0
-            };
+            var config = new MapperConfiguration(cfg => cfg.CreateMap<RelationVM, Relation>()
+            .AfterMap((src, dest) => {
+                dest.Id = Guid.NewGuid();
+                dest.EMailAddress = src.Email;
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs:199]: Remove
dest.EMailAddress = src.Email;
And test if it works.

File: MVC-App/Services/RelationService.cs [hunk 22]
```diff
@@ -200,36 +193,40 @@ public async Task<RelationListVM> GetAsync(Guid? categoryId)
...
-                PostalCode = ApplyMask(relationVM.Relation.PostalCode, _unitOfWork.CountryRepository.GetByID(relationVM.Relation.CountryId).PostalCodeFormat),
-                Number = relationVM.Relation.StreetNumber,
-                AddressTypeId = Guid.Parse("00000000-0000-0000-0000-000000000002")
-            };
+            config = new MapperConfiguration(cfg => cfg.CreateMap<RelationVM, RelationAddress>()
+            .AfterMap((src, dest) => {
+                dest.Id = Guid.NewGuid();
+                dest.RelationId = relation.Id;
+                dest.PostalCode = ApplyMask(src.PostalCode, _unitOfWork.CountryRepository.GetByID(relationVM.Relation.CountryId).PostalCodeFormat);
+                dest.Number = src.StreetNumber;
```
Comment (dmyhun-andersen) [inline MVC-App/Services/RelationService.cs:222]: Remove
dest.Number = src.StreetNumber;
And test if it works.

File: MVC-App/Services/RelationService.cs (hunk 21 above)
Comment (WMinowski) [reply MVC-App/Services/RelationService.cs:199]: it was added cause automapper fails to map this field

File: MVC-App/Services/RelationService.cs (hunk 22 above)
Comment (WMinowski) [reply MVC-App/Services/RelationService.cs:222]: same issue, mapper fails to map exact field

**Pull request merged**